promql/test.go

```diff
@@ -427,6 +428,79 @@ func (t *Test) Run() error {
 	return nil
 }
 
+type atModifierTestCase struct {
+	expr     string
+	evalTime time.Time
+}
+
+func atModifierTestCases(exprStr string, evalTime time.Time) ([]atModifierTestCase, error) {
+	expr, err := parser.ParseExpr(exprStr)
+	if err != nil {
+		return nil, err
+	}
+	ts := timestamp.FromTime(evalTime)
+
+	containsNonStepInvariant := false
+	// Setting the @ timestamp for all selectors to be evalTime.
+	// If there is a subquery, then the selectors inside it don't get the @ timestamp.
+	// If any selector already has the @ timestamp set, then it is untouched.
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		_, _, subqTs := subqueryTimes(path)
+		if subqTs != nil {
+			// There is a subquery with timestamp in the path,
+			// hence don't change any timestamp further.
+			return nil
+		}
+		switch n := node.(type) {
+		case *parser.VectorSelector:
+			if n.Timestamp == nil {
+				n.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.MatrixSelector:
+			if vs := n.VectorSelector.(*parser.VectorSelector); vs.Timestamp == nil {
+				vs.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.SubqueryExpr:
+			if n.Timestamp == nil {
+				n.Timestamp = makeInt64Pointer(ts)
+			}
+
+		case *parser.Call:
+			containsNonStepInvariant = containsNonStepInvariant || !IsFunctionStepInvariant(n)
+			if n.Func.Name == "timestamp" {
+				// If timestamp(..) has something other than vector selector inside,
```

Comment:
Are you suggesting to modify the comments here?

Reply:
I'm saying to remove this code, as it is preventing testing of things that should be tested. The comment is also wrong.

Reply:
> as it is preventing testing of things that should be tested.

What are those things?

> I'm saying to remove this code

Tests will break since timestamp() with a non instant vector will produce different results.

Reply:
That sounds like correctly detecting a bug to me.

Reply:
As discussed in -dev, I will include timestamp() in those non step invariant function map while extending it. If the above behaviour is a bug, that can be taken care in a different PR.